pennylane_lightning/lightning_kokkos/_measurements.py

```diff
@@ -0,0 +1,441 @@
+            circuit (QuantumScript): The single circuit to simulate
+            mid_measurements (None, dict): Dictionary of mid-circuit measurements
+
+        Returns:
+            Tuple[TensorLike]: The measurement results
+        """
+
+        if not circuit.shots:
+            # analytic case
+            if len(circuit.measurements) == 1:
+                return self.measurement(circuit.measurements[0])
+
+            return tuple(self.measurement(mp) for mp in circuit.measurements)
+
+        # finite-shot case
+        results = self.measure_with_samples(
+            circuit.measurements,
+            shots=circuit.shots,
+            mid_measurements=mid_measurements,
+        )
+
+        if len(circuit.measurements) == 1:
+            if circuit.shots.has_partitioned_shots:
+                return tuple(res[0] for res in results)
+
+            return results[0]
+
+        return results
+
+    def measure_with_samples(
+        self,
+        measurements: List[Union[SampleMeasurement, ClassicalShadowMP, ShadowExpvalMP]],
+        shots: Shots,
+        mid_measurements=None,
+    ) -> List[TensorLike]:
+        """
+        Returns the samples of the measurement process performed on the given state.
+        This function assumes that the user-defined wire labels in the measurement process
+        have already been mapped to integer wires used in the device.
+
+        Args:
+            measurements (List[Union[SampleMeasurement, ClassicalShadowMP, ShadowExpvalMP]]):
+                The sample measurements to perform
+            shots (Shots): The number of samples to take
+            mid_measurements (None, dict): Dictionary of mid-circuit measurements
+
+        Returns:
+            List[TensorLike[Any]]: Sample measurement results
+        """
+        # last N measurements are sampling MCMs in ``dynamic_one_shot`` execution mode
+        mps = measurements[0 : -len(mid_measurements)] if mid_measurements else measurements
+        groups, indices = _group_measurements(mps)
+
+        all_res = []
+        for group in groups:
+            if isinstance(group[0], (ExpectationMP, VarianceMP)) and isinstance(
+                group[0].obs, SparseHamiltonian
+            ):
+                raise TypeError(
+                    "ExpectationMP/VarianceMP(SparseHamiltonian) cannot be computed with samples."
+                )
+            if isinstance(group[0], VarianceMP) and isinstance(group[0].obs, (Hamiltonian, Sum)):
+                raise TypeError("VarianceMP(Hamiltonian/Sum) cannot be computed with samples.")
+            if isinstance(group[0], (ClassicalShadowMP, ShadowExpvalMP)):
+                raise TypeError(
+                    "ExpectationMP(ClassicalShadowMP, ShadowExpvalMP) cannot be computed with samples."
+                )
+            if isinstance(group[0], ExpectationMP) and isinstance(group[0].obs, Hamiltonian):
+                all_res.extend(self._measure_hamiltonian_with_samples(group, shots))
```

Comment:
This code section is not cover it by any unit test. 

Unfortunately, I could not create a test to cover it. I tried to trace the logic behind this section with the following toy code:
``` python 
import pennylane as qml

from pennylane_lightning.lightning_kokkos._measurements import (
    LightningKokkosMeasurements as LightningMeasurements,
)
from pennylane_lightning.lightning_kokkos._state_vector import (
    LightningKokkosStateVector as LightningStateVector,
)

device_name = "lightning.kokkos"
n_qubits = 4

measurement = qml.expval
observable = qml.Hamiltonian(
    [1.0, 2.0, 3.0], [qml.PauliX(0), qml.PauliY(1), qml.PauliZ(2) @ qml.PauliZ(3)]
)

ops = [qml.Hadamard(i) for i in range(n_qubits)]
measurements = [measurement(op=observable)]
tape = qml.tape.QuantumScript(ops, measurements, shots=2)

statevector = LightningStateVector(num_wires=n_qubits)
statevector = statevector.get_final_state(tape)

m = LightningMeasurements(statevector)

result = m.measure_final_state(tape)

print(f"result = {result}")
```
However, the `group[0].obs` class is different from the `Hamiltonian`.  In fact, It is decomposed into the following class: 
```
<class 'pennylane.ops.op_math.linear_combination.LinearCombination'>
```
Do you know if the function `_measure_hamiltonian_with_samples` is still necessary? @AmintorDusko @multiphaseCFD @vincentmr @maliasadi

Reply:
Maybe this is something that became obsolete? What is your opinion @PennyLaneAI/performance team?

Reply:
Therefore, I will remove the method `._measure_hamiltonian_with_samples` and this condition 
``` python
if isinstance(group[0], ExpectationMP) and isinstance(group[0].obs, Hamiltonian):
    all_res.extend(self._measure_hamiltonian_with_samples(group, shots))
```

Reply:
Nice catch! AFAIK this is an old logic that doesn't have any use cases with the new PL observable data-types. We should be able to remove it from both places.

Reply:
ok I will remove it

Reply:
Sorry folks but I discovered that the code needs these functions because if run
``` bash
pl-device-test --device lightning.kokkos --skip-ops --shots=20000
```
the test fails without them. I am not sure why this happened but I suspect that is related to legacy behaviour.